### PR TITLE
Fix incorrect pushdown of simplified domain in postgresql

### PR DESCRIPTION
**Pull request by raunaqmorarka**
For varchar/char types since Postgresql has different case
sensitivity than Trino, a simplified domain should not be
pushed down if it is not a discrete set.

### CODE REVIEW

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java [hunk 1]
```diff
@@ -233,15 +235,26 @@
...
-                domain.getValues().isDiscreteSet()) {
+        if (domain.isOnlyNull()) {
             return FULL_PUSHDOWN.apply(session, domain);
         }
 
+        // PostgreSQL is case sensitive by default
         // PostgreSQL by default orders lowercase letters before uppercase, which is different from Trino
         // TODO We could still push the predicates down if we could inject a PostgreSQL-specific syntax for selecting a collation for given comparison.
-        return DISABLE_PUSHDOWN.apply(session, domain);
+        if (!domain.getValues().isDiscreteSet()) {
```
Comment (hashhar) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:245]: Is there any way to disable pushdown for discrete sets but only when an inequality is involved? This condition is as good as turning off pushdown completely.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java [hunk 2]
```diff
@@ -233,15 +235,26 @@
...
+            return DISABLE_PUSHDOWN.apply(session, domain);
+        }
+
+        Domain simplifiedDomain = domain.simplify(getDomainCompactionThreshold(session));
+        if (!simplifiedDomain.getValues().isDiscreteSet()) {
+            // Domain#simplify can turn a discrete set into a range predicate
+            // Push down of range predicate for varchar/char types could lead to incorrect results
+            // due to different case sensitivity of PostgreSQL
```
Comment (hashhar) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:253]: super nit: due to different sort order

case-sensitivity is a boolean thing.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java (hunk 1 above)
Comment (hashhar) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:245]: I'm a bit confused now. Discrete set means independent values compared using either equality or inequality? Correct?

If so, then why will there be incorrect results? PostgreSQL is case-sensitive and so is Trino. Only the sort order is different so this should only affect inequality or ranges? Equality conditions on discrete sets should give correct results?

Comment (hashhar) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:245]: I think I understand now. A comment like ``predicate pushdown could return incorrect results for operators like `<` or `>` `` would help.

Comment (raunaqmorarka) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:245]: I've tweaked the comments a bit, hopefully it should be clearer now.
`!domain.getValues().isDiscreteSet()` should catch only range predicates

**Review by hashhar — COMMENTED** [review]
LGTM % @kokosing 's review since I'm not very familiar with the Domain concept yet.

**Review by findepi — APPROVED** [review]
Code change LGTM.

This should have been caught by a correctness test, but apparently wasn't.
Is it something we could improve?

**Comment by raunaqmorarka** [comment]
> Code change LGTM.
> 
> This should have been caught by a correctness test, but apparently wasn't.
> Is it something we could improve?

The pushed down predicate goes through PredicatePushdownController twice, once in JdbcMetadata#applyFilter and once in QueryBuilder#pushDownDomain, the 2nd pass works on simplified domain and masks the issue, preventing a correctness bug from showing up.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java [hunk 3]
```diff
@@ -291,25 +283,9 @@ public void testPredicatePushdown()
                 .isNotFullyPushedDown(
                         node(
                                 FilterNode.class,
-                                // verify that pushed down range predicate is applied by the connector
+                                // verify that no constraint is applied by the connector
```
Comment (kokosing) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:286]: add a reason why it was not applied

**Review by kokosing — APPROVED** [review]

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java (hunk 3 above)
Comment (raunaqmorarka) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:288]: done

**Pull request merged**